### PR TITLE
signed extensions: metadata type generation using reflect

### LINKED ISSUE #191: research metadata using `reflect`
Currently, the runtime metadata is being generated by `Metadata()` methods implemented in each module/extension and extrinsic. These methods are implemented by explicitly hard-coding names/types/methods/etc. We need to research whether there is a better way to generate the entire metadata. This might involve a different approach, using Go's reflect package to obtain the necessary names, types, and methods from the existing definitions in the codebase, but it might require attaching additional helper fields to the existing types.

Take a look at the `reflect` support in TinyGo. Investigate whether the implementation is sufficient to refactor the logic to be based completely on `reflect`.

**Examples**

```go
func (cg CheckGenesis) Metadata() (primitives.MetadataType, primitives.MetadataSignedExtension) {
	return primitives.NewMetadataTypeWithPath(
			metadata.CheckGenesis,
			"CheckGenesis",
			sc.Sequence[sc.Str]{"frame_system", "extensions", "check_genesis", "CheckGenesis"},
			primitives.NewMetadataTypeDefinitionComposite(sc.Sequence[primitives.MetadataTypeDefinitionField]{}),
		),
		primitives.NewMetadataSignedExtension("CheckGenesis", metadata.CheckGenesis, metadata.TypesH256)
}
```

Here, since we are in the context of `CheckGenesis` extensions there might be things that are possible to obtain through reflection: `"CheckGenesis"`, `metadata.CheckGenesis`, `primitives.NewMetadataSignedExtension("CheckGenesis", metadata.CheckGenesis, ...`.

### PR DESCRIPTION
**Pull request by nicksinch**
**Detailed description**:
- Aims to remove the necessity of writing manually a Metadata() function for each signed extension
- This is achieved using a dynamic map that stores the ids of metadata types which is initialized prior starting to build the metadata of the extrinsic
- Metadata types are generated using reflect package when iterating recursively the extensions of the signedExtra type

**Which issue(s) this PR fixes**:
Fixes: part of #191 

**Checklist**
- [ ] Documentation added
- [x] Tests updated

### CODE REVIEW

File: frame/system/extensions/check_genesis.go [hunk 1]
```diff
@@ -4,17 +4,19 @@ import (
...
 	sc "github.com/LimeChain/goscale"
-	"github.com/LimeChain/gosemble/constants/metadata"
 	"github.com/LimeChain/gosemble/frame/system"
 	primitives "github.com/LimeChain/gosemble/primitives/types"
 )
 
 type CheckGenesis struct {
-	module system.Module
+	module               system.Module
+	additionalSignedData sc.VaryingData
```
Comment (radkomih) [inline frame/system/extensions/check_genesis.go]: Could we rename it to more accurately reflect the field’s purpose? For instance, `typesInfoAdditionalSignedData` or something along those lines?

File: primitives/types/testExtraCheckComplex_test.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
+package types
```
Comment (radkomih) [inline primitives/types/testExtraCheckComplex_test.go]: file naming `testExtraCheckComplex_test.go`

File: primitives/types/extra.go [hunk 3]
```diff
@@ -142,20 +149,140 @@ func (e signedExtra) PostDispatch(pre sc.Option[sc.Sequence[Pre]], info *Dispatc
...
+	} else {
+		getExtraMetadata(extraMetadataId, metadataTypes)
+	}
+
+	constructExtension(extraValue, extraMetadataId, extensions, metadataIds, metadataTypes)
+
+	return extraMetadataId
+}
+
+func getExtraMetadata(id int, metadataTypes *sc.Sequence[MetadataType]) {
```
Comment (radkomih) [inline primitives/types/extra.go]: `appendExtraMetadata` ?

File: Makefile [hunk 4]
```diff
@@ -6,7 +6,7 @@ GC = extalloc # (extalloc, extalloc_leaking)
 VERSION = 0.31.0-dev
 IMAGE = tinygo/${TARGET}
 
-WASMOPT_PATH = tinygo/lib/binaryen/bin/wasm-opt
+WASMOPT_PATH = /tinygo/lib/binaryen/bin/wasm-opt
```
Comment (failfmi) [inline Makefile:9]: nit: unnecessary change

Comment (nicksinch) [reply Makefile:9]: I was not able to run `make build-docker-release `without it. The error was: `cannot use "tinygo/lib/binaryen/bin/wasm-opt" as wasm-opt (from WASMOPT environment variable): fork/exec tinygo/lib/binaryen/bin/wasm-opt: no such file or directoryBuild - tinygo version: 0.31.0-dev, gc: extalloc  (no debug)`

File: api/metadata/module.go [hunk 5]
```diff
@@ -208,11 +212,29 @@ func (m Module) apiMetadata() primitives.RuntimeApiMetadata {
...
+func buildMetadataTypesIdsMap() map[string]int {
+	return map[string]int{
+		"Bool":   metadata.PrimitiveTypesBool,
+		"String": metadata.PrimitiveTypesString,
+		"U8":     metadata.PrimitiveTypesU8,
+		"U16":    metadata.PrimitiveTypesU16,
+		"U32":    metadata.PrimitiveTypesU32,
+		"U64":    metadata.PrimitiveTypesU64,
+		"U128":   metadata.PrimitiveTypesU128,
+		"U256":   metadata.PrimitiveTypesU256,
```
Comment (failfmi) [inline api/metadata/module.go]: `U256` is not supported, similar to `I256`

Comment (failfmi) [reply api/metadata/module.go]: You can remove it in `func primitiveTypes()` as well

File: api/metadata/module.go [hunk 6]
```diff
@@ -105,13 +107,15 @@ func (m Module) MetadataAtVersion(dataPtr int32, dataLen int32) int64 {
 		log.Critical(err.Error())
 	}
 
+	metadataTypesIds := buildMetadataTypesIdsMap()
```
Comment (failfmi) [inline api/metadata/module.go]: The logic in `buildMetadata()` and `MetadataAtVersion` is very similar (for v14). Can we add a `TODO` to refactor it at some point later on?

File: execution/extrinsic/runtime_extrinsic_test.go [hunk 7]
```diff
@@ -13,6 +13,25 @@ import (
...
+var (
+	metadataTypesIds = map[string]int{
+		"Bool":   metadata.PrimitiveTypesBool,
+		"String": metadata.PrimitiveTypesString,
+		"U8":     metadata.PrimitiveTypesU8,
+		"U16":    metadata.PrimitiveTypesU16,
+		"U32":    metadata.PrimitiveTypesU32,
+		"U64":    metadata.PrimitiveTypesU64,
+		"U128":   metadata.PrimitiveTypesU128,
+		"U256":   metadata.PrimitiveTypesU256,
```
Comment (failfmi) [inline execution/extrinsic/runtime_extrinsic_test.go]: `U256` not supported

File: primitives/types/extra_test.go [hunk 8]
```diff
@@ -7,10 +7,32 @@ import (
...
+		"U32":    metadata.PrimitiveTypesU32,
+		"U64":    metadata.PrimitiveTypesU64,
+		"U128":   metadata.PrimitiveTypesU128,
+		"U256":   metadata.PrimitiveTypesU256,
+		"I8":     metadata.PrimitiveTypesI8,
+		"I16":    metadata.PrimitiveTypesI16,
+		"I32":    metadata.PrimitiveTypesI32,
+		"I64":    metadata.PrimitiveTypesI64,
+		"I128":   metadata.PrimitiveTypesI128,
+		"H256":   metadata.TypesH256,
```
Comment (failfmi) [inline primitives/types/extra_test.go]: Can we extract these in a static func and call it where needed? There are a few places where the map consists of the same values.

Comment (nicksinch) [reply primitives/types/extra_test.go]: I tried this but since those maps are in a **var** statement, I'm unable to add more key-values to a map after I initialize it in the 'var' statement using a static function. I get the error '<map_name> redeclared in this package'.

**Review by failfmi — APPROVED** [review]
